Add SIP source upload

jraddaoui opened the pull request:
Refs #1291.

---

**Start ingest workflow from SIP source** 

Add API endpoint to create a SIP from a SIP source and init the ingest
workflow. Add `SIPSourceID` field to the workflow request and register
two `bucketdownload` activities, one using the internal bucket and the
other using the SIP source bucket. Based on the `SIPSourceID` field run
one activity or the other.

---

**Allow SIP source upload from the dashboard**

Split the upload SIPs page into local and cloud upload using tabs. Move
the existing Uppy uploader to a component and create a new one for the
SIP source upload. In the new component, fetch a list of objects from a
hardcoded SIP source using scroll load and display a list-group with
check-boxes and multi-selection. Checking the objects and submitting
the upload will group and make all requests to the API to start an ingest
workflow for each SIP.

---

**Fix user creation on concurrent add SIP requests**

Add unique index over the user OIDC columns combined to prevent
the creation of multiple users with the same OIDC iss and sub claims.
To allow setting that index without prefix I reduced both column sizes
to 255 chars.

Additionally, when multiple requests are sent at the same time by a new
user, most of them would fail due to the new unique index. To allow that
scenario I decided to use Ent's upsert feature. This will create a new
user or update the email and name of an existing one if there is a
conflict in the OIDC columns.

djjuhasz reviewed (commented):
Wooow! It works (and looks) great so far @jraddaoui. Well done! :clap: 

I reviewed all of your commits so far and everything looked good to me, though I was mostly looking at the big picture and didn't focus on the little details.  I'll do a more thorough review when you've finished up the PR and the commits are squashed just so I'm not commenting on minor issues that may be fixed in subsequent commit.

jraddaoui committed:
Allow SIP source upload from the dashboard

Split the upload SIPs page into local and cloud upload using tabs. Move
the existing Uppy uploader to a component and create a new one for the
SIP source upload. In the new component, fetch a list of objects from a
hardcoded SIP source using scroll load and display a list-group with
check-boxes and multi-selection. Checking the objects and submitting
the upload will group and make all requests to the API to start an ingest
workflow for each SIP.

Co-authored-by: David Juhasz <djjuhasz@gmail.com>

jraddaoui commented:
Thanks @djjuhasz, a lot of that is your work and it was looking great already!

I squashed all that was done into a couple of commits and based this PR against the user creation fixes. I'll add a few more commits that we can fix-up later into the existing ones.

jraddaoui committed:
Start ingest workflow from SIP source

Add API endpoint to create a SIP from a SIP source and init the ingest
workflow. Add `SIPSourceID` field to the workflow request and register
two `bucketdownload` activities, one using the internal bucket and the
other using the SIP source bucket. Based on the `SIPSourceID` field run
one activity or the other.

Co-authored-by: David Juhasz <djjuhasz@gmail.com>

jraddaoui committed:
Fix user creation on concurrent add SIP requests

Add unique index over the user OIDC columns combined to prevent
the creation of multiple users with the same OIDC iss and sub claims.
To allow setting that index without prefix I reduced both column sizes
to 255 chars.

Additionally, when multiple requests are sent at the same time by a new
user, most of them would fail due to the new unique index. To allow that
scenario I decided to use Ent's upsert feature. This will create a new
user or update the email and name of an existing one if there is a
conflict in the OIDC columns.

jraddaoui commented:
Hi @djjuhasz, this is ready for a full look.

The idea would be to merge as three commits as outlined in the PR description. The first two and the latest one, I'll squash all the others into the dashboard commit. I thought it would be nice to leave them like that in case you want to look at some of them individually, but I could easily squash them before review if you prefer.

The latest commit relates to the user + SIP creation and the transaction PR from the other day. Sorry about the back an forward on this, in the end the `ReadOIDCUser` and `CreateUser` functions are unused for now, but I thought they could become useful in the near future.

It's still lacking some test in both ends, but that will need to happen in another PR if we want to include this in the release. BTW, I just noticed we are not getting coverage reports from Codecov, I think for quite a while already.

djjuhasz commented on dashboard/src/components/SIPUploadCloud.vue:
```diff
@@ -0,0 +1,147 @@
+<script setup lang="ts">
```
Maybe we should change the name of this component to `SIPSourceUpload.vue` to match the back-end naming and the label suggested by @fiver-watson on https://github.com/artefactual-sdps/enduro/issues/1291#issuecomment-3133525369?

If we do make  that change then I think we should change the local upload to `SIPLocalUpload` or `SIPFilesystemUpload` or something similar.

jraddaoui replied:
I used SIPUploadSource and SIPUploadLocal.

djjuhasz commented on dashboard/src/components/SIPUploadCloud.vue:
```diff
@@ -0,0 +1,147 @@
+  null,
+);
+
+useInfiniteScroll(
```
Infinite scroll, very cool! 😍

djjuhasz commented on dashboard/src/components/SIPUploadCloud.vue:
```diff
@@ -0,0 +1,153 @@
+  <div class="mb-3">
+    <div class="form-text d-flex gap-2 justify-content-end">
+      <span>Selected SIPs: {{ checkedItems.length }}</span>
+      <span>•</span>
```
It's not a blocker for this PR but I think for accessibility it would be better to use a `<menu>` tag and make the "Select all" and "Clear selections" links menu items and use CSS to add the separators between the items.

djjuhasz commented on dashboard/src/pages/ingest/upload/index.vue:
```diff
@@ -1,87 +1,64 @@
+    text: "Select from source",
+    route: router.resolve({
+      name: "/ingest/upload/",
+      query: { ...route.query, source: "cloud" },
```
Maybe "cloud" should be "source" or "sip-source"?

jraddaoui replied:
Yeah, using "source", I also changed the param name to "from" to avoid confusion.

djjuhasz commented on internal/persistence/ent/client/user.go:
```diff
@@ -97,25 +98,55 @@ func readOIDCUser(ctx context.Context, entc *db.Client, iss, sub string) (int, *
+	}
+
+	// Create or update the user. There is no save method for
+	// upsert queries so we cant get the DB user directly.
```
Typo: "can't"

djjuhasz commented on internal/persistence/ent/schema/user.go:
```diff
@@ -44,12 +44,12 @@ func (User) Fields() []ent.Field {
 		field.String("oidc_iss").
 			Annotations(entsql.Annotation{
-				Size: 1024,
+				Size: 255,
```
I'm a bit worried there could be edge cases where the issuer URL is longer than 255 chars. 🤔

jraddaoui replied:
Hopefully 255 is fine, you never know, but here an example provider URL using 255 chars generated with AI:

```
https://login.example-identity-provider.com/tenant/region/eu-west/application/authentication/v2/organizations/org-42ff83b1-91b3-4a9e-81ab-9d15fcb67a61/projects/project-4be21d8d-59a0-4d5a-8df6-3d0c9b4ac2fd/environments/production
```

We already applied that migration in the dev environment, lets leave it like that for now.

djjuhasz reviewed (approved):
This looks great to me @jraddaoui.  I suggested a few minor changes but none of them are blockers for this PR.

jraddaoui commented:
Thanks @djjuhasz! I addressed most of your feedback and, one extra thing I noticed, the route protection and all was not considering the new attributes that could allow visiting the /sips/upload page. We needed to allow OR conditions now in the route conditions so I modified the route permissions check to allow that. Also updated the sidebar check and the tabs/component checks option. Another minor detail, `onMounted` was not enough if somehow you click the upload link again while being already in that page, it removed everything from the page, we are using `watch` now ;)
